package br.com.foursys.agendamento_salas.service;

import br.com.foursys.agendamento_salas.dto.request.CriarAgendamentoRequest;
import br.com.foursys.agendamento_salas.dto.response.AgendamentoResponse;
import br.com.foursys.agendamento_salas.dto.response.ConfimacaoAgendamentoResponse;
import br.com.foursys.agendamento_salas.enums.PerfilUsuario;
import br.com.foursys.agendamento_salas.enums.StatusAgendamento;
import br.com.foursys.agendamento_salas.domain.Agendamento;
import br.com.foursys.agendamento_salas.domain.Log;
import br.com.foursys.agendamento_salas.domain.Sala;
import br.com.foursys.agendamento_salas.domain.Usuario;
import br.com.foursys.agendamento_salas.exception.*;
import br.com.foursys.agendamento_salas.mapper.AgendamentoMapper;
import br.com.foursys.agendamento_salas.port.out.AgendamentoRepositoryPort;
import br.com.foursys.agendamento_salas.port.out.LogRepositoryPort;
import br.com.foursys.agendamento_salas.port.out.SalaRepositoryPort;
import br.com.foursys.agendamento_salas.port.out.UsuarioRepositoryPort;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;


@Service
public class AgendamentoService {

    private final AgendamentoRepositoryPort agendamentoRepositoryPort;
    private final LogRepositoryPort logRepositoryPort;
    private final SalaRepositoryPort salaRepositoryPort;
    private final UsuarioRepositoryPort usuarioRepositoryPort;

    private final AgendamentoMapper agendamentoMapper;

    public AgendamentoService(AgendamentoRepositoryPort agendamentoRepositoryPort, LogRepositoryPort logRepositoryPort, SalaRepositoryPort salaRepositoryPort, UsuarioRepositoryPort usuarioRepositoryPort,
                              AgendamentoMapper agendamentoMapper) {
        this.agendamentoRepositoryPort = agendamentoRepositoryPort;
        this.logRepositoryPort = logRepositoryPort;
        this.salaRepositoryPort = salaRepositoryPort;
        this.usuarioRepositoryPort = usuarioRepositoryPort;
        this.agendamentoMapper = agendamentoMapper;
    }

    @Transactional
    public ConfimacaoAgendamentoResponse create(CriarAgendamentoRequest agendamentoRequest, Long idUsuario) {
        Usuario usuario = buscarUsuario(idUsuario);

        Sala sala = salaRepositoryPort.buscarPorId(agendamentoRequest.salaId().getId())
                .orElseThrow(SalaInexistenteException::new);

        validarHorarioAgendamento(agendamentoRequest.horaInicio(), agendamentoRequest.horaFim());

        validarQuantidadeHorasAgendamento(agendamentoRequest.horaInicio(),
                agendamentoRequest.horaFim(), usuario.getPerfilUsuario());

        validarCapacidadeMaximaSala(agendamentoRequest.qntdPessoas(), sala.getCapacidade());

        validarConflitoHorario(sala.getId(), agendamentoRequest.data(), agendamentoRequest.horaInicio(),
                agendamentoRequest.horaFim());

        if(!sala.getDisponivel()) {
            throw new SalaIndisponivelException();
        }

        Agendamento agendamento = Agendamento
                .builder()
                .usuarioId(usuario)
                .salaId(sala)
                .dataAgendamento(agendamentoRequest.data())
                .dataCriacao(LocalDateTime.now())
                .horaInicio(agendamentoRequest.horaInicio())
                .horaFim(agendamentoRequest.horaFim())
                .qntdPessoas(agendamentoRequest.qntdPessoas())
                .titulo(agendamentoRequest.titulo())
                .status(StatusAgendamento.CONFIRMADO)
                .ultimaAtualizacao(LocalDateTime.now())
                .build();

        if(agendamentoRequest.usuarioIdSolicitante() != null) {
            if(!usuario.getPerfilUsuario().equals(PerfilUsuario.ADMIN)) {
                throw new UsuarioSemPermissaoException();
            } else {
                usuarioRepositoryPort.buscarPorId(agendamentoRequest.usuarioIdSolicitante().getId())
                        .orElseThrow(UsuarioInexistenteException::new);

                agendamento.setUsuarioIdSolicitante(agendamentoRequest.usuarioIdSolicitante());
            }
        }

        agendamentoRepositoryPort.salvar(agendamento);

        Log log = Log
                .builder()
                .usuario(usuario)
                .dataAgendamento(agendamento.getDataAgendamento())
                .dataCriacao(agendamento.getDataCriacao())
                .salaAgendamento(agendamento.getSalaId())
                .horaInicioAgendamento(agendamento.getHoraInicio())
                .horaFimAgendamento(agendamento.getHoraFim())
                .build();

        logRepositoryPort.salvar(log);

        return ConfimacaoAgendamentoResponse.builder()
                .id(agendamento.getId())
                .sala(agendamento.getSalaId())
                .inicio(agendamento.getHoraInicio())
                .fim(agendamento.getHoraFim())
                .status(agendamento.getStatus())
                .solicitante(agendamento.getUsuarioIdSolicitante())
                .build();
    }

    public List<AgendamentoResponse> buscaAgendamentos(Long usuarioAutenticado) {
        buscarUsuario(usuarioAutenticado);

        List<Agendamento> listaAgendamento = agendamentoRepositoryPort.buscarUsuarioId(usuarioAutenticado);

        return listaAgendamento
                .stream()
                .map(agendamentoMapper::entityToResponse)
                .toList();
    }

    //Somente ADMIN
    public List<AgendamentoResponse> buscaTodosAgendamentos(Long usuarioAutenticado) {
        Usuario usuario = buscarUsuario(usuarioAutenticado);

        if(usuario.getPerfilUsuario() != PerfilUsuario.ADMIN) {
            throw new UsuarioSemPermissaoException();
        }

        List<Agendamento> listaAgendamento = agendamentoRepositoryPort.buscarTodos();

        return listaAgendamento
                .stream()
                .map(agendamentoMapper::entityToResponse)
                .toList();
    }

//  Métodos Utilitarios
    private long calcularDuracao(LocalTime inicio, LocalTime fim) {
        return Duration.between(inicio, fim).toMinutes();
    }

    private void validarQuantidadeHorasAgendamento(LocalTime horaInicio, LocalTime horaFim, PerfilUsuario perfilUsuario) {
        if (calcularDuracao(horaInicio, horaFim) > 120
                && perfilUsuario != PerfilUsuario.ADMIN) {
            throw new ValidacaoBuscaException(
                    "A duração não pode ser maior que 2 horas."
            );
        }
    }

    private void validarCapacidadeMaximaSala(int qtdPessoas, int capacidadeSala) {
        if(qtdPessoas > capacidadeSala) {
            throw new CapacidadeNaoSuportadaException();
        }
    }

    private void validarConflitoHorario(Long idSala, LocalDate dataAgendamento, LocalTime horaInicio, LocalTime horaFim) {
        boolean conflito = agendamentoRepositoryPort.existeConflito(idSala, dataAgendamento, horaInicio, horaFim);

        if(conflito) {
            throw new SalaIndisponivelException();
        }
    }

    private void validarHorarioAgendamento(LocalTime horaInicio, LocalTime horaFim) {
        if(horaInicio.isAfter(horaFim)) {
            throw new HorarioAgendamentoInvalidoException();
        }
    }

    private Usuario buscarUsuario(Long id) {
        return usuarioRepositoryPort.buscarPorId(id)
                .orElseThrow(UsuarioInexistenteException::new);
    }
}
